includes/razorpay-subscriptions.php

```diff
@@ -0,0 +1,349 @@
+
+        $subscriptionData = array(
+            'customer_id'     => $customerId,
+            'plan_id'         => $planId,
+            'quantity'        => (int) $product['qty'],
+            'total_count'     => $length,
+            'customer_notify' => 0,
+            'notes'           => array(
+                'woocommerce_order_id'   => $orderId,
+                'woocommerce_product_id' => $product['product_id']
+            ),
+        );
+
+        $signUpFee = WC_Subscriptions_Product::get_sign_up_fee($product['product_id']);
+
+        if ($signUpFee)
+        {
+            $item = array(
+                'amount'   => (int) round($signUpFee * 100),
+                'currency' => get_woocommerce_currency(),
+                'name'     => $product['name']
+            );
+
+            if ($item['currency'] !== self::INR)
+            {
+                $this->razorpay->handleCurrencyConversion($item);
+            }
+
+            $subscriptionData['addons']['item'] = $item;
+        }
+
+        return $subscriptionData;
+    }
+
+    protected function getProductPlanId($product)
+    {
+        $productId = $product['product_id'];
+
+        $metadata = get_post_meta($productId);
+
+        // Creating a new plan only if plan isn't created already
+        // Or if the plan has changed
+        if ($this->shouldCreatePlan($metadata, $productId) === true)
+        {
+            $planId = $this->createPlan($product);
+
+            add_post_meta($productId, self::RAZORPAY_PLAN_ID, $planId, true);
+        }
+        else
+        {
+            // Retrieve the plan id if already created
+            $planId = $metadata[self::RAZORPAY_PLAN_ID][0];
+        }
+
+        return $planId;
+    }
+
+    /**
+     * We shouldn't create plan only if the following conditions are met
+     * 1. There exists a plan id as product metadata
+     * 2. The corresponding plan has amount a1
+     * 3. The sale price is amount a2
+     * 4. a1 is not equal to a2, therefore the plan has changed
+     *
+     * @param $metadata
+     * @param $productId
+     */
+    protected function shouldCreatePlan($metadata, $productId)
+    {
+        if (empty($metadata[self::RAZORPAY_PLAN_ID]) === false)
+        {
+            $planId = $metadata[self::RAZORPAY_PLAN_ID][0];
+
+            try
+            {
+                $plan = $this->api->plan->fetch($planId);
+            }
+            catch (Exception $e)
+            {
+                echo "Razorpay Error: " . $e->getMessage();
+            }
+
+            //
+            // If the product sale price is different from the plan amount
+            // Or if the product price is different from the plan amount
+            // We delete the old plan id from post metadata.
+            // Then, we create a new plan and save id as metadata
+            //
+            if (($this->ifProductOnSale($plan, $productId) === true) or
```

Comment:
I don't understand. Why won't this work:

1. Try to create a new plan and get the arguments as per the current state. This will obviously include whether or not there is a sale and what is the current price
2. Match it against whatever plan we already have on file.

If they don't match, create a new plan. Else use the old one.